package/cfg/k3s-cis-1.9/policies.yaml

```diff
@@ -0,0 +1,380 @@
+        remediation: |
+          Identify all clusterrolebindings to the cluster-admin role. Check if they are used and
+          if they need this role or if they could use a role with fewer privileges. K3s gives exceptions
+          to the helm-kube-system-traefik and helm-kube-system-traefik-crd clusterrolebindings
+          as these are required for traefik installation into the kube-system namespace for regular operations.
+          Where possible, first bind users to a lower privileged role and then remove the
+          clusterrolebinding to the cluster-admin role:
+          ```
+          kubectl delete clusterrolebinding [name]
+          ```
+        scored: false
+
+      - id: 5.1.2
+        text: "Minimize access to secrets (Automated)"
+        audit: "echo \"canGetListWatchSecretsAsSystemAuthenticated: $(kubectl auth can-i get,list,watch secrets --all-namespaces --as=system:authenticated)\""
+        tests:
+          test_items:
+            - flag: "canGetListWatchSecretsAsSystemAuthenticated"
+              compare:
+                op: eq
+                value: no
+        remediation: |
+          Where possible, remove get, list and watch access to Secret objects in the cluster.
+        scored: true
+
+      - id: 5.1.3
+        text: "Minimize wildcard use in Roles and ClusterRoles (Manual)"
+        audit: |
+          # Check Roles
+          kubectl get roles --all-namespaces -o custom-columns=ROLE_NAMESPACE:.metadata.namespace,ROLE_NAME:.metadata.name --no-headers | while read -r role_namespace role_name
+          do
+            role_rules=$(kubectl get role -n "${role_namespace}" "${role_name}" -o=json | jq -c '.rules')
+            if echo "${role_rules}" | grep -q "\[\"\*\"\]"; then
+              printf "**role_name: %-50s  role_namespace: %-25s role_rules: %s role_is_compliant: false\n" "${role_name}" "${role_namespace}" "${role_rules}"
+            else
+              printf "**role_name: %-50s role_namespace: %-25s role_is_compliant: true\n" "${role_name}" "${role_namespace}"
+            fi;
+          done
+
+          # Check ClusterRoles
+          kubectl get clusterroles -o custom-columns=CLUSTERROLE_NAME:.metadata.name --no-headers | while read -r clusterrole_name
+          do
+            clusterrole_rules=$(kubectl get clusterrole "${clusterrole_name}" -o=json | jq -c '.rules')
+            if echo "${clusterrole_rules}" | grep -q "\[\"\*\"\]"; then
+              echo "**clusterrole_name: ${clusterrole_name} clusterrole_rules: ${clusterrole_rules} clusterrole_is_compliant: false"
+            else
+              printf "**clusterrole_name: %-50s clusterrole_is_compliant: true\n" "${clusterrole_name}"
+            fi;
+          done
+        remediation: |
+          Where possible replace any use of wildcards in clusterroles and roles with specific
+          objects or actions.
+        scored: false
```

Comment:
Are there any Roles and ClusterRoles with "*" in K3s that are required, and that we could whitelist too ? (Same as 5.1.1 approach).

Reply:
I am unable to completely fix this check due to limitations on how kube-bench processes it checks. The `or` operation will not work, on the failure detection for the first operator, the process is restarted **ON ALL LINES** for the second operation. This effectively makes every operation an implicit and.

Example given the following lines of output:
```
clusterrole_name: admin clusterrole_is_compliant: true
clusterrole_name: clustercidrs-node clusterrole_is_compliant: true
clusterrole_name: edit clusterrole_is_compliant: true
clusterrole_name: k3s-cloud-controller-manager clusterrole_is_compliant: false
```
Assume a check of `is_compliant: true` OR `clusterrole_name: k3s-cloud-controller-manager`:

The first check for `is_compliant` will fail on the 4th line. This causes the process to move onto the next check "is the clusterrole_name == k3s-cloud-controller-manger" but it restarts on the top line, not on the 4th line that failed. So the second check will fail on the first line, because `admin != k3s-cloud-controller-manager`.

Reply:
From what I recall, it was one of the issue I stumbled upon when writing the CIS-1.9 checks. I initially had split this check into two distinct checks, but then had to follow the CIS check numbering, as advised by kube-bench's folks.

In this situation, since we can't break the check down into 2 checks, we could either:
-  improve kube-bench engine OR
- improve the bash script, to only use `is_compliant` for both `roles` and `clusterroles`. Similar to 5.1.6, and use the valid_elements as you did.

See below:

```
     - id: 5.1.3
         text: "Minimize wildcard use in Roles and ClusterRoles (Manual)"
         audit: |
           # Check Roles
           kubectl get roles --all-namespaces -o custom-columns=ROLE_NAMESPACE:.metadata.namespace,ROLE_NAME:.metadata.name --no-headers | while read -r role_namespace role_name
           do
             role_rules=$(kubectl get role -n "${role_namespace}" "${role_name}" -o=json | jq -c '.rules')
             if echo "${role_rules}" | grep -q "\[\"\*\"\]"; then
               printf "**role_name: %-50s  role_namespace: %-25s role_rules: %s is_compliant: false\n" "${role_name}" "${role_namespace}" "${role_rules}"
             else
               printf "**role_name: %-50s role_namespace: %-25s is_compliant: true\n" "${role_name}" "${role_namespace}"
             fi;
           done
           # Check ClusterRoles
           kubectl get clusterroles -o custom-columns=CLUSTERROLE_NAME:.metadata.name --no-headers | while read -r clusterrole_name
           do
             clusterrole_rules=$(kubectl get clusterrole "${clusterrole_name}" -o=json | jq -c '.rules')
             if echo "${clusterrole_rules}" | grep -q "\[\"\*\"\]"; then
               echo "**clusterrole_name: ${clusterrole_name} clusterrole_rules: ${clusterrole_rules}  is_compliant: false"
             else
               printf "**clusterrole_name: %-50s is_compliant: true\n" "${clusterrole_name}"
             fi;
           done
         use_multiple_values: true
         tests:
           bin_op: or
           test_items:
             - flag: "is_compliant"
               compare:
                 op: eq
                 value: true
             - flag: "service_account"
               compare:
                 op: valid_elements
                 value: ...,... <- put the roles and cluster roles to whitelist
```

WDYT ?